components/skus/browser/skus_service.cc

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/skus/browser/skus_service.h"
+
+#include <utility>
+
+#include "base/json/json_reader.h"
+#include "base/task/post_task.h"
+#include "base/threading/sequenced_task_runner_handle.h"
+#include "brave/components/skus/browser/pref_names.h"
+#include "brave/components/skus/browser/rs/cxx/src/lib.rs.h"
+#include "brave/components/skus/browser/rs/cxx/src/shim.h"
+#include "brave/components/skus/browser/skus_context_impl.h"
+#include "brave/components/skus/browser/skus_utils.h"
+#include "components/prefs/pref_service.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+
+namespace {
+
+void OnRefreshOrder(skus::RefreshOrderCallbackState* callback_state,
+                    skus::SkusResult result,
+                    rust::cxxbridge1::Str order) {
+  std::string order_str = static_cast<std::string>(order);
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run(order_str);
+  }
+  delete callback_state;
+}
+
+void OnFetchOrderCredentials(
+    skus::FetchOrderCredentialsCallbackState* callback_state,
+    skus::SkusResult result) {
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run("");
+  }
+  delete callback_state;
+}
+
+void OnPrepareCredentialsPresentation(
+    skus::PrepareCredentialsPresentationCallbackState* callback_state,
+    skus::SkusResult result,
+    rust::cxxbridge1::Str presentation) {
+  if (callback_state->cb) {
+    std::move(callback_state->cb).Run(static_cast<std::string>(presentation));
+  }
+  delete callback_state;
+}
+
+void OnCredentialSummary(skus::CredentialSummaryCallbackState* callback_state,
+                         skus::SkusResult result,
+                         rust::cxxbridge1::Str summary) {
+  std::string summary_string = static_cast<std::string>(summary);
+  base::JSONReader::ValueWithError value_with_error =
+      base::JSONReader::ReadAndReturnValueWithError(
+          summary_string, base::JSONParserOptions::JSON_PARSE_RFC);
+  absl::optional<base::Value>& records_v = value_with_error.value;
+
+  if (records_v && callback_state->prefs) {
+    if (callback_state->domain == "vpn.brave.com" ||
+        callback_state->domain == "vpn.bravesoftware.com" ||
+        callback_state->domain == "vpn.brave.software") {
+      const base::Value* active = records_v->FindKey("active");
+      if (active) {
+        bool has_credential = active && active->is_bool() && active->GetBool();
+        callback_state->prefs->SetBoolean(skus::prefs::kSkusVPNHasCredential,
```

Comment:
It doesn't make any sense to pass through prefs because there's no reason for this to be an anonymous method. If there was a potential lifetime issue (there isn't because SkusService owns the rust sdk) then this would be totally unsafe anyway.

Reply:
I can make `OnCredentialSummary` a member method, but it would complicate things (`brave/components/skus/browser/rs/cxx/src/shim.h` defines this callback as `base::OnceCallback<void(const std::string&)>`). Will open group DM about this and the above comment

Reply:
Fixed in https://github.com/brave/brave-core/pull/10847/commits/918ee0a11f6e9b1653d7edde1a713d3eca98a9f8